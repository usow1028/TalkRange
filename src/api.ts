import express, { NextFunction, Request, Response } from 'express';
import { DEFAULT_CULTURE, DEFAULT_PROFILE } from './config.js';
import { ev } from './ev.js';
import { infer } from './inference.js';
import {
  ActionRec,
  Profile,
  RangeRequestBody,
  RangeResponse,
} from './types.js';

/**
 * Clamp numeric value to the [0,1] interval.
 * 값을 0~1 범위로 제한.
 */
const clamp01 = (value: number): number => Math.min(1, Math.max(0, value));

/**
 * Merge incoming profile overrides with defaults.
 * 입력 프로필을 기본값과 병합.
 */
const sanitizeProfile = (incoming?: Partial<Profile>): Profile => {
  if (!incoming) {
    return DEFAULT_PROFILE;
  }
  return {
    relationship: (incoming.relationship ?? DEFAULT_PROFILE.relationship) as Profile['relationship'],
    taskUrgency: clamp01(incoming.taskUrgency ?? DEFAULT_PROFILE.taskUrgency),
    futureImportance: clamp01(incoming.futureImportance ?? DEFAULT_PROFILE.futureImportance),
    tolerance: clamp01(incoming.tolerance ?? DEFAULT_PROFILE.tolerance),
  };
};

/**
 * Validate and sanitize conversation history array.
 * 대화 히스토리 배열을 검증/정제.
 */
const ensureHistory = (history?: string[]): string[] => {
  if (!history) return [];
  return history.filter((item): item is string => typeof item === 'string');
};

/**
 * Validate REST payload and return error message if invalid.
 * REST 본문을 검증하고 문제 시 오류 메시지 반환.
 */
const validateBody = (body: RangeRequestBody): string | undefined => {
  if (!body.role) return 'role is required';
  if (!body.time_context) return 'time_context is required';
  if (!body.utterance) return 'utterance is required';
  return undefined;
};

/**
 * Construct the TalkRange express application with configured routes.
 * TalkRange Express 애플리케이션을 생성한다.
 */
export const createApp = (): express.Express => {
  const app = express();
  app.use(express.json());

  app.get('/health', (_req: Request, res: Response) => {
    res.status(200).json({ status: 'ok' });
  });

  app.post('/range', (req: Request, res: Response) => {
    const body = req.body as RangeRequestBody;
    const error = validateBody(body);
    if (error) {
      res.status(400).json({ error });
      return;
    }

    const culture = body.culture ?? DEFAULT_CULTURE;
    const history = ensureHistory(body.history);
    const profile = sanitizeProfile(body.my_profile);

    const result = infer(body.role, body.time_context, culture, body.utterance, history);
    const actions = ev(result.range, profile, culture, body.role);

    const response: RangeResponse = {
      intent_range: result.range,
      recommended_actions: actions.map((item: ActionRec) => ({
        ...item,
        intents: item.intents.map((intent) => ({ ...intent })),
      })),
      explain: {
        signals: result.signals,
        note: result.note,
      },
    };

    res.json(response);
  });

  app.use((error: Error, _req: Request, res: Response, _next: NextFunction) => {
    void _next;
    // eslint-disable-next-line no-console
    console.error('TalkRange API error', error);
    res.status(500).json({ error: 'internal_error' });
  });

  return app;
};
