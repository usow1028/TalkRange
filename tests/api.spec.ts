import request from 'supertest';
import scenarios from '../data/scenarios.example.json';
import { createApp } from '../src/api.js';

describe('TalkRange API', () => {
  const app = createApp();

  it('responds to health check', async () => {
    const response = await request(app).get('/health');
    expect(response.status).toBe(200);
    expect(response.body).toEqual({ status: 'ok' });
  });

  it('validates request body', async () => {
    const response = await request(app).post('/range').send({
      time_context: 'now',
      utterance: 'hello',
    });
    expect(response.status).toBe(400);
    expect(response.body).toHaveProperty('error');
  });

  it('returns range and actions for scenario sample', async () => {
    const scenario = (scenarios as unknown as Array<Record<string, unknown>>)[0];

    const response = await request(app)
      .post('/range')
      .send({
        role: scenario.role,
        time_context: scenario.time_context,
        culture: scenario.culture,
        history: scenario.history,
        utterance: scenario.utterance,
        my_profile: scenario.my_profile,
      });

    expect(response.status).toBe(200);
    expect(Array.isArray(response.body.intent_range)).toBe(true);
    expect(Array.isArray(response.body.recommended_actions)).toBe(true);

    const probabilityTotal = response.body.intent_range.reduce(
      (acc: number, item: { probability: number }) => acc + item.probability,
      0,
    );
    expect(probabilityTotal).toBeCloseTo(1, 5);

    expect(response.body).toMatchSnapshot({
      intent_range: expect.arrayContaining([
        expect.objectContaining({
          intent: expect.any(String),
          probability: expect.any(Number),
        }),
      ]),
      recommended_actions: expect.arrayContaining([
        expect.objectContaining({
          action: expect.any(String),
          ev: expect.any(Number),
          rationale: expect.any(String),
          template: expect.any(String),
        }),
      ]),
      explain: {
        signals: expect.any(Array),
        note: expect.any(String),
      },
    });
  });
});
