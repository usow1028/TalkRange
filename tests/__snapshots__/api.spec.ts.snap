// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TalkRange API returns range and actions for scenario sample 1`] = `
{
  "explain": {
    "note": Any<String>,
    "signals": Any<Array>,
  },
  "intent_range": ArrayContaining [
    ObjectContaining {
      "intent": Any<String>,
      "probability": Any<Number>,
    },
  ],
  "recommended_actions": ArrayContaining [
    ObjectContaining {
      "action": Any<String>,
      "ev": Any<Number>,
      "rationale": Any<String>,
      "template": Any<String>,
    },
  ],
}
`;
